tools/asset-inventory/asset_inventory/bigquery_schema.py

```diff
@@ -0,0 +1,339 @@
+    if not first_character.isalpha() and first_character != '_':
+        new_property_name = '_' + new_property_name
+    new_property_name = new_property_name[:128]
+
+    # did the property name change?
+    if property_name != new_property_name:
+        property_value = parent.pop(property_name)
+        parent[new_property_name] = property_value
+
+    # handle labels (condition #1)
+    if new_property_name == 'labels':
+        _convert_labels_dict_to_list(parent)
+
+    property_value = parent[new_property_name]
+
+    # recursivly descend.
+    sanitized = sanitize_property_value(property_value, depth=depth + 1)
+
+    # else the value could have changed.
+    parent[new_property_name] = sanitized
+
+    # remove empty dicts or list of empty dicts (condition #3)
+    if is_empty_dict_list_or_empty_dict(sanitized):
+        # BigQuery doesn't deal well with empty records.
+        # prune the value.
+        parent.pop(new_property_name)
+
+
+def is_empty_dict_list_or_empty_dict(property_value):
+    """True if is an empty dict or a list of empty dicts."""
+
+    if isinstance(property_value, dict) and not property_value:
+        return True
+
+    if isinstance(property_value, list):
+        for property_element in property_value:
+            if ((not isinstance(property_element, dict)) or property_element):
+                return False
+        return True
+    return False
+
+
+def sanitize_property_value(property_value, depth=0):
+    """Modifies supplied json object for BigQuery load.
+
+    Traverses the json object and modifies it to conform to BigQuery
+    requirements.
+
+    1. Will prune any value with more then 15 layers of nesting as that's the
+    most BigQuery will handle. See `_sanitize_property` for description of other
+    rules enforced.
+
+    2. rounds/truncates numeric values to be between BigQuery limits for NUMERIC
+    values.
+
+    Args:
+        property_value: Json object.
+        depth: Level of embedding within the json document.
+    Returns:
+        Modified json object.
+
+    """
+
+    # BigQuery can't deal with more then 15 nested fields.
+    # prune it.
+    if depth > 15:
+        return {}
+
+    # NUMERIC data type is an exact numeric value with 38 digits of precision
+    # and 9 decimal digits of scale.
+    if isinstance(property_value, Number):
+        if isinstance(property_value, float):
+            property_value = round(property_value, 9)
+        property_value = max(property_value, MIN_NUMERIC)
+        property_value = min(property_value, MAX_NUMERIC)
+
+    # sanitize each nested list element.
+    if isinstance(property_value, list):
+        for array_item in property_value:
+            sanitize_property_value(array_item, depth)
+
+    # and each nested json object.
+    if isinstance(property_value, dict):
+        for child_property in dict(property_value):
+            _sanitize_property(child_property, property_value, depth)
```

Comment:
_sanitize_property does more then clean up the value.   it can also change the name of the property, or delete the property entirely.  I could have it return the new property name, the new property value and a flag if the property should be deleted but that sounds like it doesn't make anything less complex.